Diagnostics landing (#13744)
dotnet/docs: docs/core/diagnostics/logging-tracing.md

---
title: .NET Core Logging and Tracing
description: An introduction to .NET Core Logging and Tracing.
author: sdmaclea
ms.author: stmaclea
ms.date: 08/05/2019
---

> # .NET Core Logging and Tracing

Reviewer: ```suggestion
# .NET Core logging and tracing
```

> Logging and tracing is a primitive technique used since the early days of computers. It simply involves instrumenting an application to produce output to be consumed later.

Author: Reworked this slightly differently

> Logging and tracing are similar. As used here we'll distinguish:

Reviewer: We should be avoided unless Microsoft is recommending something
https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/w/we

```suggestion
Logging and tracing are similar. In this article, these two terms are distinguished as follows:

```

- Logging outputs text as human readable content.

> - Tracing outputs binary content.  It can use tokenizing, and filtering to reduce overhead in the traced application.

Reviewer: ```suggestion
- Tracing outputs binary content. It can use tokenizing and filtering to reduce overhead in the traced application.
```

Reviewer: tokenizing seems weird. Is that correct or should it be tokenization?

> ## Why Logging and Tracing?

Reviewer: ```suggestion
## Reasons to use logging and tracing
```

These simple techniques are surprisingly powerful. They can be used in situations where a fancy debugger fails:

- Issues occurring over long periods of time, can be difficult to debug with a traditional debugger.
- Multi-threaded applications and distributed applications are often difficult to debug.  Attaching a debugger tends to modify behaviors. Detailed logs can be analyzed as needed to understand complex systems.
- Many services shouldn't be stalled. Attaching a debugger often causes timeout failures.

## Logging APIs

> <xref:System.Console?displayProperty=nameWithType>, <xref:System.Diagnostics.Trace?displayProperty=nameWithType>, and <xref:System.Diagnostics.Debug?displayProperty=nameWithType> each provide similar APIs convenient for logging.

Reviewer: ```suggestion
The <xref:System.Console?displayProperty=nameWithType>, <xref:System.Diagnostics.Trace?displayProperty=nameWithType>, and <xref:System.Diagnostics.Debug?displayProperty=nameWithType> classes provide similar APIs convenient for logging.
```

> The choice of which API to use is up to you. The key differences:

Reviewer: ```suggestion
The choice of which API to use is up to you. The key differences are:
```

- <xref:System.Console?displayProperty=nameWithType>

>   - Always compiled in and always writes to the console.

Reviewer: what does compiled in means?

>   - Useful for information that your customer may need to see in release.

Reviewer: ```suggestion
  - Useful for information that your customer may need to see in the release.
```

  - Because it's the simplest approach, it's often used for ad-hoc temporary debugging. This debug code is often never checked in to source control.
- <xref:System.Diagnostics.Trace?displayProperty=nameWithType>
  - Only enabled when `TRACE` is defined.

>   - Writes to attached <xref:System.Diagnostics.TraceListener>s. Usually a debugger.

Reviewer 2: I'm not sure what the "Usually a debugger" means?

Reviewer: Don't make the plural by attaching the s because it's bad for loc. Define what the TraceListener is. Object?

```suggestion
  - Writes to attached <xref:System.Diagnostics.TraceListener> objects. Usually a debugger.
```

>   - It seems best to use this API when creating release-oriented logs.

Reviewer 2: I'm not sure how to interpret 'release-oriented'? Were you trying to call attention to the fact that DEBUG is usually only defined debug builds vs. TRACE is usually defined all builds?

>   - It seems best to use this API when creating release-oriented logs.

Reviewer: It seems best is a weird recommendation. can we be more direct?

- <xref:System.Diagnostics.Debug?displayProperty=nameWithType>
  - Only enabled when `DEBUG` is defined.

>   - Writes to attached <xref:System.Diagnostics.TraceListener>s. Usually a debugger.

Reviewer: Ditto

>   - It seems best to use this API when creating debug-oriented logs.

Reviewer: Ditto

> ### Logging Related References

Reviewer: ```suggestion
### Logging related references
```

> - [How to: Compile Conditionally with Trace and Debug](../../framework/debug-trace-profile/how-to-compile-conditionally-with-trace-and-debug)

Reviewer: ```suggestion
- [How to: Compile Conditionally with Trace and Debug](../../framework/debug-trace-profile/how-to-compile-conditionally-with-trace-and-debug.md)
```

> - [How to: Add Trace Statements to Application Code](../../framework/debug-trace-profile/how-to-add-trace-statements-to-application-code)

Reviewer: ```suggestion
- [How to: Add Trace Statements to Application Code](../../framework/debug-trace-profile/how-to-add-trace-statements-to-application-code.md)
```

- [ASP.NET Logging](/aspnet/core/fundamentals/logging)
provides an overview of the logging techniques it supports.

> - [C# String Interpolation](../../csharp/language-reference/tokens/interpolated)

Reviewer: ```suggestion
- [C# String Interpolation](../../csharp/language-reference/tokens/interpolated.md)
```

 can simplify writing logging code.

> - The <xref:System.Exception.Message?displayProperty=nameWithType> is useful for logging exceptions.

Reviewer: ```suggestion
- The <xref:System.Exception.Message?displayProperty=nameWithType> property is useful for logging exceptions.
```

> - The <xref:System.Diagnostics.StackTrace?displayProperty=nameWithType>

Reviewer: ```suggestion
- The <xref:System.Diagnostics.StackTrace?displayProperty=nameWithType> class
```

> can be useful to provide debugger like logging info.

Reviewer: ```suggestion
can be useful to provide logging information to the debugger.
```

- <xref:System.Diagnostics.TraceListener> provides an overview of existing listeners and creating a custom listener.

## Tracing APIs

- <xref:System.Diagnostics.Tracing.EventSource?displayProperty=nameWithType>

>   - EventSource is the primary root .NET Core tracing API

Reviewer: ```suggestion
  - EventSource is the primary root .NET Core tracing API.
```

>   - Available in all .NET Standard versions

Reviewer: ```suggestion
  - Available in all .NET Standard versions.
```

>   - Emits to .NET Core's event pipe on all platforms

Reviewer: ```suggestion
  - Emits to .NET Core's event pipe on all platforms.
```

>   - Also emits to ETW on Windows.

Reviewer: Yes, we need to spell out ETW. Link also helps.

>   - Only allows tracing serializable objects

Reviewer: ```suggestion
  - Only allows tracing serializable objects.
```

- <xref:System.Diagnostics.DiagnosticSource?displayProperty=nameWithType>

>   - Only available in .NET Core

Reviewer: ```suggestion
  - Only available in .NET Core.
```

>   - Allows tracing non-serializable objects

Reviewer: ```suggestion
  - Allows tracing non-serializable objects.
```

  - Includes a bridge to allow select fields of logged objects to be written to an <xref:System.Diagnostics.Tracing.EventSource>.

- <xref:System.Diagnostics.TraceSource?displayProperty=nameWithType>

>   - This API was introduced in .NET Framework 2.0

Reviewer: ```suggestion
  - This API was introduced in .NET Framework 2.0.
```

Reviewer: Is this information relevant since this is a .NET Core article?

>   - Supported in .NET Standard 2.0

Reviewer: ```suggestion
  - Supported in .NET Standard 2.0.
```

>   - Prefer <xref:System.Diagnostics.Tracing.EventSource>

Reviewer: ```suggestion
  - Prefer <xref:System.Diagnostics.Tracing.EventSource> instead.
```

Reviewer: Do we want to say why?

Author: On Noah's suggestion, I removed the `TraceSource` reference.

- <xref:System.Diagnostics.EventLog?displayProperty=nameWithType>

>   - Windows only

Reviewer: ```suggestion
  - Windows only.
```

>   - Writes messages to Windows event log

Reviewer: ```suggestion
  - Writes messages to the Windows event log.
```

  - System administrators expect fatal application error messages to appear in the Windows event log.

> ## Performance Considerations

Reviewer 2: I'm wondering if we need this section or if we should put it somewhere else? Most of the page described options available for logging and then here we are switching to some best practices for authoring logging.

Reviewer: ```suggestion
## Performance considerations
```

String formatting can take noticeable CPU processing time.

> In performance critical applications, it's good to:

Reviewer: ```suggestion
In performance critical applications, it's recommended that you:
```

> - Avoid lots of logging when no one is listening

Reviewer: ```suggestion
- Avoid lots of logging when no one is listening.
```

> - Only log what is useful

Reviewer: ```suggestion
- Only log what's useful.
```

> - Defer fancy formatting to the analysis stage

Reviewer: ```suggestion
- Defer fancy formatting to the analysis stage.
```

- Many services shouldn't be stalled. Attaching a debugger often causes timeout failures.
- Issues aren't always foreseen. Logging and tracing are designed for low overhead so that programs can always be recording in case an issue occurs.

## .NET Core APIs

<xref:System.Console?displayProperty=nameWithType>, <xref:System.Diagnostics.Trace?displayProperty=nameWithType>, and <xref:System.Diagnostics.Debug?displayProperty=nameWithType> each provide similar print style APIs convenient for logging.

The choice of which print style API to use is up to you. The key differences:
- <xref:System.Console?displayProperty=nameWithType>
  - Always enabled and always writes to the console.
  - Useful for information that your customer may need to see in the release.
  - Because it's the simplest approach, it's often used for ad-hoc temporary debugging. This debug code is often never checked in to source control.
- <xref:System.Diagnostics.Trace?displayProperty=nameWithType>
  - Only enabled when `TRACE` is defined.
  - Writes to attached <xref:System.Diagnostics.Trace.Listeners>, by default the <xref:System.Diagnostics.DefaultTraceListener>.

>   - It seems best to use this API when creating logs that will be enabled in most builds.

Reviewer 3: Not really the type of language I'd expect to see in a guidance doc. As we are open to other vendors in this space, you can just say it's the preferred inbox solution for logging in production.

- <xref:System.Diagnostics.Debug?displayProperty=nameWithType>
  - Only enabled when `DEBUG` is defined.
  - Writes to an attached debugger.
  - On `*nix` writes to stderr if `COMPlus_DebugWriteToStdErr` is set.

>   - It seems best to use this API when creating logs that will only be useful in a debguenabled in debug builds.

Reviewer 3: Just as last comment, seems is a little vague. Maybe say: This API is useful to provide additional logging that's useful only if a [managed debugger](md link here) is attached. Also typo in `debguenabled`, although not sure if you mean `that will only be useful in debug-enabled builds`?

- <xref:System.Diagnostics.Tracing.EventSource?displayProperty=nameWithType>
  - EventSource is the primary root .NET Core tracing API.
  - Available in all .NET Standard versions.
  - Only allows tracing serializable objects.
  - Writes to attached <xref:System.Diagnostics.Tracing.EventListener>s.
  - .NET Core provides listeners for:

>     - .NET Core's event pipe on all platforms

Reviewer 4: I'd prefer to reword `event pipe`  to `EventPipe` for the sake of consistency across documentations.

    - [Event Tracing for Windows (ETW)](/windows/win32/etw/event-tracing-portal)
    - [LTTng tracing framework for Linux](https://lttng.org/)

- <xref:System.Diagnostics.DiagnosticSource?displayProperty=nameWithType>
  - Included in .NET Core and as a [Nuget package](https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource) for .NET Framework
  - Allows in process tracing non-serializable objects.
  - Includes a bridge to allow select fields of logged objects to be written to an <xref:System.Diagnostics.Tracing.EventSource>.

- <xref:System.Diagnostics.Activity?displayProperty=nameWithType>
  - Provides a definitive way to identify log messages resulting from a specific activity or transaction. This object can be used to correlate logs across different services.

- <xref:System.Diagnostics.EventLog?displayProperty=nameWithType>
  - Windows only.

>   - Writes messages to Windows event log.

Reviewer 4: nit: I think `Windows Event Log` is the name they use in Windows API docs so we might want to capitalize those E and L to match with their docs: https://docs.microsoft.com/en-us/windows/win32/wes/windows-event-log

>   - System administrators expect fatal application error messages to appear in the Windows event log.

Reviewer 4: same as above
